pilot/pkg/model/service.go

```diff
@@ -586,15 +586,15 @@ func ParseServiceKey(s string) (hostname host.Name, ports PortList, lc labels.Co
 // BuildSubsetKey generates a unique string referencing service instances for a given service name, a subset and a port.
 // The proxy queries Pilot with this key to obtain the list of instances in a subset.
 func BuildSubsetKey(direction TrafficDirection, subsetName string, hostname host.Name, port int) string {
-	return fmt.Sprintf("%s|%d|%s|%s", direction, port, subsetName, hostname)
```

Comment:
@howardjohn From the beanchmark, there is a big improve for const str. I would argue we revert this , which is more readable.

Reply:
if there is a big improvement why should we revert it?

Reply:
```
$ go test -run=^$ -bench=BenchmarkBuildSubsetKey ./pilot/pkg/model
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/model
BenchmarkBuildSubsetKey-8       27351206                45.1 ns/op
PASS
ok      istio.io/istio/pilot/pkg/model  2.205s
gke ~/go/src/istio.io/istio (perf/sprintf)
$ go test -run=^$ -bench=BenchmarkBuildSubsetKey ./pilot/pkg/model
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/model
BenchmarkBuildSubsetKey-8        3192273               355 ns/op
PASS
ok      istio.io/istio/pilot/pkg/model  1.660s
```

it is 8x faster. We call this method 1000000s of times, so it adds up pretty quickly. This one method was like 1% of pilot CPU

Reply:
i am surprised fmt.sprintf is slower compared to doing the same ops using string conversion etc. 🤷‍♂

Reply:
at the very least, can you leave the old sprintf commented out so that people know the format they are dealing with?

Reply:
I think it does a bunch of allocs and parsing for the formats and junk like that

Reply:
I would agree with this, but could you declare const separator for both "|" and "_."

Reply:
I don't know why the article showed such a big change, but in our case there is no difference. Probably `"|"` is treated the same as a const and the compiler optimizes it.

See
```
$ go test ./pilot/pkg/model -test.run=^$ -bench=BenchmarkBuildSubsetKey -benchmem -test.benchtime=10s
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/model
BenchmarkBuildSubsetKey-6       308374704               38.1 ns/op             0 B/op          0 allocs/op
PASS
ok      istio.io/istio/pilot/pkg/model  15.833s
kind-kind ~/go/src/istio.io/istio (perf/sprintf)
$ go test ./pilot/pkg/model -test.run=^$ -bench=BenchmarkBuildSubsetKey -benchmem -test.benchtime=10s
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/model
BenchmarkBuildSubsetKey-6       311559710               38.1 ns/op             0 B/op          0 allocs/op
PASS
ok      istio.io/istio/pilot/pkg/model  15.903
```